### ISSUE TITLE
feat: remove requirement for `PLUGIN_TYPE` positional arg in CLI commands `meltano add`, `meltano remove`, `meltano install`

### ISSUE THREAD

**Issue by tayloramurphy** [meltano/meltano#7066]
Currently in the CLI only the `meltano add` and `meltano remove` commands require you to specify the plugin type. 

Other commands default to just referencing the name while enabling a user to specify `--plugin-type` if necessary. Commands that accept `--plugin-type`:

* [invoke](https://docs.meltano.com/reference/command-line-interface#invoke)
* [lock](https://docs.meltano.com/reference/command-line-interface#lock)
* [config](https://docs.meltano.com/reference/command-line-interface#config)

**Proposal**

Drop the requirement for <type> when using meltano add and remove so that users can type:

`meltano add tap-gitlab` or `meltano remove target-postgres` and Meltano can just figure that out. When there is a collision a we can error out, tell the user there are multiple plugins with that name present and have them disambiguate.

When a user is adding a `--custom` plugin, we can infer the type from the name (pattern match on `tap` and `target`) and/or prompt them for a type. We should likely just ask is this an extractor, loader, or a non-connector (utility).

Related:

* https://github.com/meltano/meltano/issues/6866
* https://github.com/meltano/meltano/issues/3263

**Comment by tayloramurphy** [comment]
cc @sbalnojan @afolson @DouweM 

Likely a nice developer joy win to simplify building a meltano project.

**Comment by edgarrmondragon** [comment]
> and/or prompt them for a type. We should likely just ask is this an extractor, loader, or a non-connector (utility).

My vote is for this in custom plugins 👍

**Comment by edgarrmondragon** [comment]
Also, `meltano add <plugin>` may be inefficient with the current Hub API model and a new JSON endpoint/file may need to be created to avoid looping through all plugin types in the client.

**Comment by tayloramurphy** [comment]
> Also, meltano add <plugin> may be inefficient with the current Hub API model and a new JSON endpoint/file may need to be created to avoid looping through all plugin types in the client.

@edgarrmondragon I hadn't thought about that, but seems eminently solvable 😄

**Comment by aaronsteers** [comment]
> Also, `meltano add <plugin>` may be inefficient with the current Hub API model and a new JSON endpoint/file may need to be created to avoid looping through all plugin types in the client.

What about the top-level index here:? https://hub.meltano.com/meltano/api/v1/plugins/index

It includes a `ref` for all variants, as well as the default variant.

We'd have to page through the types, but locally that iteration would be pretty quick.

**Comment by aaronsteers** [comment]
@tayloramurphy - Is your goal with this issue to address `meltano add` and `meltano remove`, or also `meltano install`?

Scope of `4` may be accurate for `add` and `remove`. I don't know about `meltano install` - probably that would need to have separate scope.

**Comment by edgarrmondragon** [comment]
> What about the top-level index here:? [hub.meltano.com/meltano/api/v1/plugins/index](https://hub.meltano.com/meltano/api/v1/plugins/index)
> 
> It includes a `ref` for all variants, as well as the name of the default variant.
> 
> We'd have to page through the types, but locally that iteration would be pretty quick.

@aaronsteers yeah that'd probably be faster, if still pulling more data than it needs.

**Comment by aaronsteers** [comment]
> > What about the top-level index here:? [hub.meltano.com/meltano/api/v1/plugins/index](https://hub.meltano.com/meltano/api/v1/plugins/index)
> > It includes a `ref` for all variants, as well as the name of the default variant.
> > We'd have to page through the types, but locally that iteration would be pretty quick.
> 
> @aaronsteers yeah that'd probably be faster, if still pulling more data than it needs.

Thanks for confirming. Agreed it's perhaps slightly more than needed, but at least not an order of magnitude greater. The future optimized version would likely take some search parameters like 'plugin name' and perhaps offer pagination if the resultset is too large. But that said, not being static, I don't know if the optimized version would be much faster even if sending less data. 🤷 (We can tackle+discover when we get there. 😄 )

**Comment by edgarrmondragon** [comment]
@aaronsteers With just `meltano install` and `meltano remove` this could be an easy `4`.

With `meltano add` this probably becomes an `8`, or even `12`, since it's more complex (Hub request, API changes to `ProjectAddService` and `ProjectPlugin`, etc.)

**Comment by tayloramurphy** [comment]
@aaronsteers my preference / priority list on this would be:

* add
* remove
* install

I mostly care about `add` as it's part of the onboarding and getting started flow. Happy to pull `install` out of scope for this particular issue.

**Comment by WillDaSilva** [comment]
Still relevant

**Comment by tayloramurphy** [comment]
I'd still love this @edgarrmondragon. Big lift but a great UX win

**Comment by tayloramurphy** [comment]
@edgarrmondragon was this supposed to be closed by #9351 ?

**Comment by edgarrmondragon** [comment]
> [@edgarrmondragon](https://github.com/edgarrmondragon) was this supposed to be closed by [#9351](https://github.com/meltano/meltano/pull/9351) ?

Yup!

**Comment by tayloramurphy** [comment]
@edgarrmondragon nice work! Exciting to see this closed after all these years 😄
